soc/nordic/common/pm_s2ram.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/arch/cpu.h>
+#include <zephyr/linker/sections.h>
+#include <zephyr/arch/common/pm_s2ram.h>
+#include <hal/nrf_resetinfo.h>
+#include <pm_s2ram.h>
+
+
+#define NVIC_MEMBER_SIZE(member) ARRAY_SIZE(((NVIC_Type *)0)->member)
+
+/* Currently dynamic regions are only used in case of userspace or stack guard and
+ * stack guard is not used by default on Cortex-M33 because there is a dedicated
+ * mechanism for stack overflow detection. Unless those condition change we don't
+ * need to store MPU content, it can just be reinitialized on resuming.
+ */
+#define MPU_USE_DYNAMIC_REGIONS IS_ENABLED(CONFIG_USERSPACE) || IS_ENABLED(CONFIG_MPU_STACK_GUARD)
+
+/* TODO: The num-mpu-regions property should be used. Needs to be added to dts bindings. */
+#define MPU_MAX_NUM_REGIONS 16
```

Comment:
you could do one thing here: have a static variable, and the first time you run the code, set it to `ARM_MPU_TYPE()`, other runs could skip this read if variable != 0.